### DISCUSSION TITLE
HasModelSchema Trait Upgrade: Removing getFillable for Better Validation in Laravel

### DISCUSSION THREAD

**Discussion by ntiyiso-rikhotso** [laravel/framework#47952] [Ideas]
Laravel, being a popular PHP framework, continues to evolve with powerful new features. In this update, I am thrilled to introduce the `ModelSchema` attribute, designed to simplify and streamline data validation for models.

Imagine a scenario where you have a model named `User` or `Product`, and you want to define specific database fields and their validation rules without using the traditional `fillable` or `guarded` properties. The `ModelSchema` attribute is the perfect solution for this.

Let me illustrate this feature using the example of a `User` model:

```php
namespace App\Models;

use App\Attributes\ModelSchema;

#[ModelSchema([
    'name' => 'required|string|max:255',
    'email' => 'required|email|unique:test_users,email',
    'age' => 'integer|min:18',
    'gender' => 'required|in:male,female,other',
    'is_active' => 'boolean',
])]
class User extends Model
{
    public $timestamps = false;
    protected $guarded = [];
}
```

In this example, the `ModelSchema` attribute lets you conveniently define the attributes and their validation rules directly in the model, using a clean and expressive syntax. The attributes are listed along with their respective validation constraints, such as `required`, `string`, `email`, `integer`, `min`, `in`, and `boolean`. 

This approach simplifies the validation process and ensures that only valid data is stored in the database. Gone are the days of manually defining the `fillable` or `guarded` properties and dealing with complex validation logic in your controllers.

Now, validating incoming data becomes seamless using the model's schema. For instance, in a controller, you can utilize the `validate` method to automatically validate incoming requests against the defined schema rules:

```php
namespace App\Http\Controllers;

use App\Models\User;
use Illuminate\Http\Request;

class UsersController extends Controller
{
    public function store(Request $request)
    {
          $user = new User([
          'name' => "Taylor",
          'email' => "taylor@laravel.com",
          'age' => 40,
          'gender' => 'male',
       ]);

       // You can modify the validation rules for this specific instance if needed
       $user->validate(function ($rules) {
          // Custom validation logic
           $rules['age'] = 'integer|min:30';
           return $rules;
        });
        
        // Or
        
        $user->validate();
        
        if($user->passes()){
          // save
        }
    }
}
```

With the `ModelSchema` attribute, you achieve cleaner code, maintain a separation of concerns, and experience a more streamlined development process. Embrace this feature to enjoy the benefits of simplified data validation in your Laravel models!

**Comment by henzeb** [comment]
Is this already implemented? I don't recall seeing an PR for this. If not... I am not a big fan of its name. Modelschema sounds more like it's defining the schema of the table, rather than specifying validation... 

Also, it should not replace fillables and guards. Rather complement them. As there are enough use cases where this would block things and you may have fields that you did not specify as you don't need them which might get sent to you anyway.

**Comment by ntiyiso-rikhotso** [comment]
@henzeb thanks for the input, this is a work in progress locally. You are more than welcome to suggest something that you think might be the best suited for this proposal. Another name that I think might work is `FieldsDecorator` and can be used as follows:
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Attributes\FieldsDecorator;

#[FieldsDecorator([
    'name' => 'required|string|max:255',
    'email' => 'required|email|unique:test_users,email', 
    'age' => 'integer|min:18',
    'gender' => 'required|in:male,female,other',
    'is_active' => 'boolean',
])]
class User extends Model
{
    public $table = 'test_users';
    public $timestamps = false;
    protected $guarded = [];
}
```

and you could do something like:

```php
$user = new User;
$user->name = "Taylor";
$user->is_active = true;

$user->validate();

// Or

$user->validate(fn($rules) => []); // to remove all rules

if($user->isValid()){
  // perform other actions
}
```
I think this might also complement other attributes like `$casts`, `$fillable`, `$guarded`
